### ISSUE TITLE
Add a setting to enforce strict type declarations for empty collection initialization

### ISSUE THREAD

**Issue by SimonSchick** [microsoft/pyright#4075]
**Is your feature request related to a problem? Please describe.**
*Generally* collections are pretty homogenous in terms of their content, thus it would be advantageous to enforce that any empty collection initialization have strict types at to ensure no incorrect values are added to it later which can lead to bugs.

**Describe the solution you'd like**

Add a setting called `reportUntypedEmptyCollectionInitialization` which will flag all of:

```py
a_list = []
a_set = set()
a_dict = {}
```

```py
a_list: List[MyType] = []
a_set = set[MyType]()
a_dict: Dict[MyType, MyType2] = {}
```

Some of those scenarios (especially sets and list operations) are implicitly covered:

<img width="614" alt="image" src="https://user-images.githubusercontent.com/3259089/197243444-2a278dd0-88cc-4def-8a3d-d4c72e5b84a0.png">

The same should probably apply to any [collection](https://docs.python.org/3/library/collections.html) whose generic argument cannot be inferred by its initializer.

In fact the scope of this may be expanded to *any* language construct/class/function call etc. whose generic argument cannot be inferred, this might be a much bigger lift though.

**Additional context**
I don't think so?
I'm unsure if this is by design in general, this seems like a pretty big oversight, couldn't find anything in the issues.

**Comment by erictraut** [comment]
Can you provide a concrete example of where the current strict-mode rules do not report a problem that leads to a potential hole in type checking — i.e. a case where an actual runtime bug would go undetected by the static type checker?

**Comment by erictraut** [comment]
Pyright tracks all "implicit Any" types as a special form called `Unknown`. In strict type checking mode (e.g. when rules like `reportUnknownVariableType` are enabled), it is an error to use an unknown or partially-unknown type. Those usages represent potential "holes" in type checking. It is not considered an error to generate an `Unknown` type because that is, by itself, is not indicative of a typing "hole".

In your examples above, the `reportUnknownVariableType` rule is correctly identifying the use of a variable with a partially-unknown type. An additional `reportUntypedEmptyCollectionInitialization` check would redundant and wouldn't catch any error that is not already caught with the existing rules. It would also be inconsistent with the general principles that we've applied elsewhere in pyright.

If you can provide a concrete example of where the current strict-mode rules leave a potential hole in type checking, please let us know. We can then consider how to plug that hole in a manner that is consistent with the established type checking principles in pyright.
